File: frontend/src/types/erp-document.ts
export type ErpDocumentType =
  | "TECH_PACK"
  | "BOM"
  | "PURCHASE_ORDER"
  | "GRN"
  | "INSPECTION_REPORT"
  | "LAB_TEST_REPORT"
  | "PRODUCTION_ORDER"
  | "STOCK_TRANSFER"
  | "INVENTORY_ADJUSTMENT"
  | "PICK_LIST"
  | "PACKING_LIST"
  | "SHIPPING_MARKS"
  | "ASN"
  | "COMMERCIAL_INVOICE"
  | "OUTGOING_INVOICE"
  | "BILL_OF_LADING"
  | "CERTIFICATE_OF_ORIGIN"
  | "EXPORT_DECLARATION"
  | "LETTER_OF_CREDIT"
  | "BILL_OF_EXCHANGE"
  | "PROOF_OF_DELIVERY"
  | "PAYMENT_RECORD"
  | "LANDED_COST"

export type ErpDocumentStatus = "DRAFT" | "ISSUED" | "CONFIRMED" | "CANCELLED"

export type JourneyStep = {
  document_type: ErpDocumentType
  journey_step: number
  phase: string
  label: string
  slug: string
  number_prefix: string
}

export type JourneyResponse = {
  steps: JourneyStep[]
  phases: string[]
}

export type ErpDocument = {
  id: number
  document_number: string
  document_type: ErpDocumentType
  type_label: string
  journey_step: number
  phase: string
  status: ErpDocumentStatus
  title: string
  reference_number: string | null
  notes: string | null
  content: Record<string, unknown>
  supplier_id: number | null
  customer_id: number | null
  purchase_id: number | null
  related_document_id: number | null
  created_by_id: number | null
  created_at: string
  updated_at: string
}

export type ErpDocumentListResponse = {
  items: ErpDocument[]
  total: number
}

export type ErpDocumentJourneyGroup = {
  phase: string
  journey_step_start: number
  steps: JourneyStep[]
  documents: ErpDocument[]
}

export type ErpDocumentJourneyViewResponse = {
  phases: ErpDocumentJourneyGroup[]
  total_documents: number
}

File: backend/app/schemas/erp_document.py
from datetime import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ErpDocumentStatus, ErpDocumentType


class JourneyStepRead(BaseModel):
    document_type: ErpDocumentType
    journey_step: int
    phase: str
    label: str
    slug: str
    number_prefix: str


class JourneyResponse(BaseModel):
    steps: list[JourneyStepRead]
    phases: list[str]


class ErpDocumentBase(BaseModel):
    document_type: ErpDocumentType
    title: str = Field(min_length=1, max_length=255)
    reference_number: str | None = Field(default=None, max_length=128)
    notes: str | None = None
    content: dict[str, Any] = Field(default_factory=dict)
    supplier_id: int | None = None
    customer_id: int | None = None
    purchase_id: int | None = None
    sale_id: int | None = None
    related_document_id: int | None = None


class ErpDocumentCreate(ErpDocumentBase):
    status: ErpDocumentStatus = ErpDocumentStatus.DRAFT


class ErpDocumentUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    status: ErpDocumentStatus | None = None
    reference_number: str | None = Field(default=None, max_length=128)
    notes: str | None = None
    content: dict[str, Any] | None = None
    supplier_id: int | None = None
    customer_id: int | None = None
    purchase_id: int | None = None
    sale_id: int | None = None
    related_document_id: int | None = None


class ErpDocumentRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_number: str
    document_type: ErpDocumentType
    type_label: str
    journey_step: int
    phase: str
    status: ErpDocumentStatus
    title: str
    reference_number: str | None
    notes: str | None
    content: dict[str, Any]
    supplier_id: int | None
    customer_id: int | None
    purchase_id: int | None
    sale_id: int | None
    related_document_id: int | None
    created_by_id: int | None
    created_at: datetime
    updated_at: datetime


class ErpDocumentListResponse(BaseModel):
    items: list[ErpDocumentRead]
    total: int


class ErpDocumentJourneyGroup(BaseModel):
    phase: str
    journey_step_start: int
    steps: list[JourneyStepRead]
    documents: list[ErpDocumentRead]


class ErpDocumentJourneyViewResponse(BaseModel):
    phases: list[ErpDocumentJourneyGroup]
    total_documents: int
